File: src/prototype/web/tests/unit/control_layer.test.ts
// Tests for ControlLayer (initiative) — docs/design/state_machines_v1.md §7.

import { describe, expect, it } from "vitest";
import {
  INITIAL_CONTROL_LAYER,
  updateControlLayer,
  type ControlLayerInputs,
} from "../../src/state/control_layer.js";
import { INITIAL_JUDGMENT_WINDOW } from "../../src/state/judgment_window.js";
import { initialActorState, type ActorState } from "../../src/state/game_state.js";
import type { HandFSM } from "../../src/state/hand_fsm.js";

function gripped(side: "L" | "R"): HandFSM {
  return Object.freeze({
    side,
    state: "GRIPPED" as const,
    target: "SLEEVE_R" as const,
    stateEnteredMs: 0,
    reachDurationMs: 0,
    lastParriedZone: null,
    lastParriedAtMs: Number.NEGATIVE_INFINITY,
  });
}

function actor(over: Partial<ActorState> = {}): ActorState {
  return Object.freeze({ ...initialActorState(0), ...over });
}

function inputs(over: Partial<ControlLayerInputs> = {}): ControlLayerInputs {
  return {
    judgmentWindow: INITIAL_JUDGMENT_WINDOW,
    bottom: actor(),
    top: actor(),
    defenderCutInProgress: false,
    ...over,
  };
}

describe("ControlLayer rule order (§7.2)", () => {
  it("judgment window OPEN locks initiative to the firing side", () => {
    const win = Object.freeze({
      ...INITIAL_JUDGMENT_WINDOW,
      state: "OPEN" as const,
      firedBy: "Bottom" as const,
      stateEnteredMs: 0,
      candidates: Object.freeze(["SCISSOR_SWEEP" as const]),
    });
    const out = updateControlLayer(INITIAL_CONTROL_LAYER, inputs({ judgmentWindow: win }));
    expect(out.initiative).toBe("Bottom");
    expect(out.lockedByWindow).toBe(true);
  });

  it("arm_extracted on bottom → Bottom", () => {
    const out = updateControlLayer(INITIAL_CONTROL_LAYER, inputs({
      bottom: actor({ armExtractedLeft: true }),
    }));
    expect(out.initiative).toBe("Bottom");
  });

  it("arm_extracted on top → Top", () => {
    const out = updateControlLayer(INITIAL_CONTROL_LAYER, inputs({
      top: actor({ armExtractedRight: true }),
    }));
    expect(out.initiative).toBe("Top");
  });

  it("bottom with ≥2 GRIPPED hands → Bottom", () => {
    const out = updateControlLayer(INITIAL_CONTROL_LAYER, inputs({
      bottom: actor({ leftHand: gripped("L"), rightHand: gripped("R") }),
    }));
    expect(out.initiative).toBe("Bottom");
  });

  it("defender cut in progress → Top", () => {
    const out = updateControlLayer(INITIAL_CONTROL_LAYER, inputs({
      defenderCutInProgress: true,
    }));
    expect(out.initiative).toBe("Top");
  });

  it("nothing applies → Neutral", () => {
    const out = updateControlLayer(INITIAL_CONTROL_LAYER, inputs());
    expect(out.initiative).toBe("Neutral");
    expect(out.lockedByWindow).toBe(false);
  });

  it("window lock releases when window returns to CLOSED", () => {
    const lockedPrev = Object.freeze({ initiative: "Bottom" as const, lockedByWindow: true });
    const out = updateControlLayer(lockedPrev, inputs());
    expect(out.lockedByWindow).toBe(false);
    expect(out.initiative).toBe("Neutral");
  });
});

File: src/prototype/unity/Assets/BJJSimulator/Tests/EditMode/ControlLayerTest.cs
// NUnit EditMode mirror of src/prototype/web/tests/unit/control_layer.test.ts.
// Each [Test] corresponds to one it(...) case from the Stage 1 Vitest suite.
//
// Run from Unity Editor: Window → General → Test Runner → EditMode.

using NUnit.Framework;

namespace BJJSimulator.Tests
{
    [TestFixture]
    public class ControlLayerTest
    {
        static HandFSM GrippedHand(HandSide side) => new HandFSM
        {
            Side = side, State = HandState.Gripped, Target = GripZone.SleeveR,
            StateEnteredMs = 0, ReachDurationMs = 0,
            LastParriedZone = GripZone.None, LastParriedAtMs = BJJConst.SentinelTimeMs,
        };

        static ActorState Actor(
            bool armExtL = false, bool armExtR = false,
            HandFSM? lh = null, HandFSM? rh = null)
        {
            var a = GameStateOps.InitialActorState();
            return new ActorState
            {
                LeftHand          = lh ?? a.LeftHand,
                RightHand         = rh ?? a.RightHand,
                LeftFoot          = a.LeftFoot,
                RightFoot         = a.RightFoot,
                PostureBreak      = Vec2.Zero,
                Stamina           = 1f,
                ArmExtractedLeft  = armExtL,
                ArmExtractedRight = armExtR,
            };
        }

        static ControlLayerInputs Inputs(
            JudgmentWindow? win = null,
            ActorState? bottom = null, ActorState? top = null,
            bool defenderCut = false)
        {
            return new ControlLayerInputs
            {
                JudgmentWindow        = win    ?? JudgmentWindow.Initial,
                Bottom                = bottom ?? GameStateOps.InitialActorState(),
                Top                   = top    ?? GameStateOps.InitialActorState(),
                DefenderCutInProgress = defenderCut,
            };
        }

        // -----------------------------------------------------------------------

        [Test]
        public void JudgmentWindow_OPEN_Locks_Initiative_To_FiredSide_Bottom()
        {
            var win = new JudgmentWindow
            {
                State           = JudgmentWindowState.Open,
                StateEnteredMs  = 0,
                Candidates      = new[] { Technique.ScissorSweep },
                CooldownUntilMs = BJJConst.SentinelTimeMs,
                FiredBy         = WindowSide.Bottom,
            };
            var result = ControlLayerOps.Update(ControlLayer.Initial, Inputs(win: win));
            Assert.AreEqual(Initiative.Bottom, result.Initiative);
            Assert.IsTrue(result.LockedByWindow);
        }

        [Test]
        public void ArmExtracted_On_Bottom_Gives_Bottom_Initiative()
        {
            var result = ControlLayerOps.Update(ControlLayer.Initial,
                Inputs(bottom: Actor(armExtL: true)));
            Assert.AreEqual(Initiative.Bottom, result.Initiative);
        }

        [Test]
        public void ArmExtracted_On_Top_Gives_Top_Initiative()
        {
            var result = ControlLayerOps.Update(ControlLayer.Initial,
                Inputs(top: Actor(armExtR: true)));
            Assert.AreEqual(Initiative.Top, result.Initiative);
        }

        [Test]
        public void Bottom_With_TwoGrips_Gives_Bottom_Initiative()
        {
            var result = ControlLayerOps.Update(ControlLayer.Initial, Inputs(
                bottom: Actor(lh: GrippedHand(HandSide.L), rh: GrippedHand(HandSide.R))));
            Assert.AreEqual(Initiative.Bottom, result.Initiative);
        }

        [Test]
        public void DefenderCutInProgress_Gives_Top_Initiative()
        {
            var result = ControlLayerOps.Update(ControlLayer.Initial, Inputs(defenderCut: true));
            Assert.AreEqual(Initiative.Top, result.Initiative);
        }

        [Test]
        public void Nothing_Applies_Gives_Neutral()
        {
            var result = ControlLayerOps.Update(ControlLayer.Initial, Inputs());
            Assert.AreEqual(Initiative.Neutral, result.Initiative);
            Assert.IsFalse(result.LockedByWindow);
        }

        [Test]
        public void WindowLock_Releases_When_Window_Returns_To_Closed()
        {
            var locked = new ControlLayer { Initiative = Initiative.Bottom, LockedByWindow = true };
            var result = ControlLayerOps.Update(locked, Inputs());
            Assert.IsFalse(result.LockedByWindow);
            Assert.AreEqual(Initiative.Neutral, result.Initiative);
        }
    }
}
